core/src/main/java/io/grpc/ManagedChannelBuilder.java

```diff
@@ -75,6 +75,19 @@
   }
 
   /**
+   * Execute application code directly in the transport thread.
+   *
+   * <p>Depending on the underlying transport, using a direct executor may lead to substantial
+   * performance improvements. However, it also requires the application to not block under
+   * any circumstances.
+   *
+   * <p>Calling this method is semantically equivalent to calling {@link #executor(Executor)} and
+   * passing in a direct executor. However, this is the preferred way as it may allow the transport
+   * to perform special optimizations.
+   */
+  public abstract T directExecutor();
```

Comment:
@louiscryan @ejona86 

yeah that's true. I guess the question is how much room for error do we want to leave for the user, as we don't have a way of checking if an executor is single threaded. What I like about `directExecutor()` is that one can't do anything wrong. If you use it, you get a speedboost, else you don't BUT it's still correct.

> Alternatives to consider
> 
> withExecutorType(ExecutorTypeEnum { DEFAULT, DIRECT, SINGLE-THREADED, ...})

I suppose this could make sense, if we can make the case for why a user would want to pass in his own single-threaded Executor. Can't really think of a good one?

> withThreads(int numthreads) - GRPC creates the executor itself, 0 implies direct etc.

That sounds good to me. Maybe a different name though? Can't think of a good one myself. `applicationThreads(), applicationParallelism(), concurrencyLevel(), withParallelism(), applicationConcurrency()`?